timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/GetSuspectedLeaderWithUuid.java

```diff
@@ -0,0 +1,150 @@
+class GetSuspectedLeaderWithUuid implements Consumer<List<BatchElement<UUID, Optional<ClientAwarePingableLeader>>>> {
+
+    private static final Logger log = LoggerFactory.getLogger(GetSuspectedLeaderWithUuid.class);
+
+    private final Map<ClientAwarePingableLeader, ExecutorService> executors;
+    private final UUID localUuid;
+    private final Duration leaderPingResponseWait;
+
+    private final Map<UUID, ClientAwarePingableLeader> cache = Maps.newHashMap();
+
+    GetSuspectedLeaderWithUuid(
+            Map<ClientAwarePingableLeader, ExecutorService> executors,
+            UUID localUuid,
+            Duration leaderPingResponseWait) {
+        this.executors = executors;
+        this.localUuid = localUuid;
+        this.leaderPingResponseWait = leaderPingResponseWait;
+    }
+
+    @Override
+    public void accept(List<BatchElement<UUID, Optional<ClientAwarePingableLeader>>> batchElements) {
+        Multimap<UUID, SettableFuture<Optional<ClientAwarePingableLeader>>> uuidsToRequests = batchElements.stream()
+                .collect(ImmutableListMultimap.toImmutableListMultimap(BatchElement::argument, BatchElement::result));
+
+        KeyedStream.of(uuidsToRequests.keySet())
+                .filterKeys(cache::containsKey)
+                .map(cache::get)
+                .forEach((cachedUuid, pingable) -> completeRequest(uuidsToRequests, cachedUuid, Optional.of(pingable)));
+
+        Set<UUID> uncachedUuids = uuidsToRequests.keySet().stream()
+                .filter(uuid -> !cache.containsKey(uuid))
+                .collect(toSet());
+
+        if (uncachedUuids.isEmpty()) {
+            return;
+        }
+
+        PaxosResponsesWithRemote<ClientAwarePingableLeader, PaxosContainer<UUID>> results =
+                PaxosQuorumChecker.collectUntil(
+                ImmutableList.copyOf(executors.keySet()),
+                pingable -> PaxosContainer.of(pingable.uuid()),
+                executors,
+                leaderPingResponseWait,
+                state -> state.responses().values().stream().map(PaxosContainer::get).collect(toSet())
+                        .containsAll(uncachedUuids));
+
+        for (Map.Entry<ClientAwarePingableLeader, PaxosContainer<UUID>> resultEntries : results.responses().entrySet()) {
+            ClientAwarePingableLeader pingable = resultEntries.getKey();
+            UUID uuid = resultEntries.getValue().get();
+
+            ClientAwarePingableLeader oldCachedEntry = cache.putIfAbsent(uuid, pingable);
+            throwIfInvalidSetup(oldCachedEntry, pingable, uuid);
+            completeRequest(uuidsToRequests, uuid, Optional.of(pingable));
+        }
+
+        Set<UUID> missingUuids = Sets.difference(
+                uncachedUuids,
+                results.withoutRemotes().stream().map(PaxosContainer::get).collect(toSet()));
+
+        missingUuids.forEach(missingUuid -> completeRequest(uuidsToRequests, missingUuid, Optional.empty()));
+    }
+
+    private static void completeRequest(
```

Comment:
This should just take a `Collection<SettableFuture<Optional<ClientAwarePingableLeader>>>`